File: stalker_pyramid/static/stalker/js/Studio.js
// Stalker Pyramid
// Copyright (C) 2013 Erkan Ozgur Yilmaz
//
// This file is part of Stalker Pyramid.
//
// This library is free software; you can redistribute it and/or
// modify it under the terms of the GNU Lesser General Public
// License as published by the Free Software Foundation;
// version 2.1 of the License.
//
// This library is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public
// License along with this library; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA

try {
    var moment = require('../../moment/moment.min');
} catch (e) {};


function Studio() {
    'use strict';
    this.daily_working_hours = 9;
    this.weekly_working_hours = 51;
    this.working_hours = { // these should be in UTC time/hours
        1: [28800000, 61200000],  // monday    9 : 10:00 - 19:00 in Turkey 08:00 - 17:00 in UTC
        2: [28800000, 61200000],  // tuesday   9 : 10:00 - 19:00 in Turkey 08:00 - 17:00 in UTC
        3: [28800000, 61200000],  // wednesday 9 : 10:00 - 19:00 in Turkey 08:00 - 17:00 in UTC
        4: [28800000, 61200000],  // thursday  9 : 10:00 - 19:00 in Turkey 08:00 - 17:00 in UTC
        5: [28800000, 61200000],  // friday    9 : 10:00 - 19:00 in Turkey 08:00 - 17:00 in UTC
        6: [36000000, 57600000],  // saturday  6 : 12:00 - 18:00 in Turkey 10:00 - 16:00 in UTC
        7: [null, null]           // sunday    0 : day off
    };
    this.vacations = [];
}

/**
 * Returns the total working hours between the given dates
 * 
 * @param {Date, Number} start
 * @param {Date, Number} end
 */
Studio.prototype.get_working_hours_between_dates = function (start, end) {
    'use strict';
    // calculate the working hours between the given dates as milliseconds
    // don't consider vacations for now

    // so we need to start from start and iterate over to the end date
    // each time by setting our temp date to the start working hour and end
    // working hour of that day

    var temp_start = +start,
        temp_end = +start,
        end_date = +end,
        total_working_millies = 0;
    while (temp_start < end_date) {
        total_working_millies += temp_end - temp_start;
        temp_start = this.get_closest_working_hour_start(temp_end);
        temp_end = this.get_closest_working_hour_end(temp_start);
        temp_end = Math.min(temp_end, end_date);
        temp_start = Math.min(temp_start, temp_end);
    }
    return total_working_millies;
};

/**
 * Returns the next closest working hour start if the given date is not in a
 * working hour range. Otherwise returns the date itself.
 * 
 * @param {Date} date
 *      Integer or Date object
 */
Studio.prototype.get_closest_working_hour_start = function (date) {
    'use strict';
    // ensure it is a Date instance
    // because it is the start value, to prevent the last millisecond of the
    // working day to be counted as a working hour we add 1 milliseconds to the
    // start value to correctly get the next working days start value as the
    // return value
    var is_working_hour = this.is_working_hour(date + 1);

    var temp_date = +date;

    if (is_working_hour) {
        return temp_date;
    } else {
        // get the day and the working hours for that day, if no working hours
        // go to the next day
        temp_date = moment.utc(date);
        var day_of_week = temp_date.isoWeekday();
        // now get the working hours for that day
        var working_hours = this.working_hours[day_of_week];
        // if date.hour is lower than start return start
        // else check if the next day is a working day
        var clock = (temp_date.valueOf() % 86400000);
        if (working_hours[0] !== null && clock <= working_hours[0]) {
            return temp_date.startOf('day') + working_hours[0];
        } else {
            // we should check the next day
            // because of the time offsets there is very weird things going on
            // when daylight saving is active,
            // and as a shortcut directly set to the next days working hours start
            var next_day_of_week = temp_date.add(1, 'day').isoWeekday();
            var next_day_working_hour_start = this.working_hours[next_day_of_week][0] || 0;

            temp_date = temp_date.startOf('day') + next_day_working_hour_start;
            return +this.get_closest_working_hour_start(temp_date);
        }
    }
};

/**
 * Returns the next closest working hour end if the given date is not in a
 * working hour range. Otherwise returns the date itself.
 * 
 * @param {Date} date
 *      Integer or Date object
 */
Studio.prototype.get_closest_working_hour_end = function (date) {
    'use strict';
    // ensure it is a Date instance
    var is_working_hour = this.is_working_hour(date);

    // get the day and the working hours for that day, if no working hours
    // go to the next day
    var temp_date = moment.utc(date);
    var day_of_week = temp_date.isoWeekday();
    // now get the working hours for that day
    var working_hours = this.working_hours[day_of_week];
    // if date.hour is lower than end return end
    // else check if the next day is a working day
    var clock = (temp_date.valueOf() % 86400000);
    if (working_hours[1] !== null && clock <= working_hours[1]) {
        return temp_date.startOf('day') + working_hours[1];
    } else {
        // we should check the next day
        temp_date.add(1, 'day').startOf('day');
        return +this.get_closest_working_hour_end(+temp_date);
    }
};


/**
 * Returns true or false depending on if the given date is in a working
 * hour range
 *   
 * @param {Date} date
 * @returns {boolean}
 */
Studio.prototype.is_working_hour = function (date) {
    'use strict';
    // get the working hours of the week day
    var week_day, clock, working_hours_of_week_day, temp_date;
    // ensure that it is a Date object

    var temp_date = moment.utc(date);

    week_day = temp_date.isoWeekday();
    clock = temp_date - moment.utc(temp_date).startOf('day');

//    clock = temp_date.getClock();
    working_hours_of_week_day = this.working_hours[week_day];
    return working_hours_of_week_day[0] !== null &&
        working_hours_of_week_day[1] !== null &&
        clock >= working_hours_of_week_day[0] &&
        clock <= working_hours_of_week_day[1];
};

try {
    module.exports.Studio = Studio;
} catch (e) {
}

File: stalker_pyramid/views/studio.py
# -*- coding: utf-8 -*-
# Stalker Pyramid a Web Base Production Asset Management System
# Copyright (C) 2009-2014 Erkan Ozgur Yilmaz
#
# This file is part of Stalker Pyramid.
#
# This library is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation;
# version 2.1 of the License.
#
# This library is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with this library; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA

import logging

from pyramid.httpexceptions import HTTPOk
from pyramid.view import view_config

from stalker.db import DBSession
from stalker import Studio, WorkingHours
from stalker_pyramid.views import (get_time, PermissionChecker)

logger = logging.getLogger(__name__)
logger.setLevel(logging.WARNING)

#
# @view_config(
#     route_name='dialog_create_studio',
#     renderer='templates/studio/dialog_create_studio.jinja2'
# )
# def create_studio_dialog(request):
#     """creates the content of the create_studio_dialog
#     """
#     return {
#         'mode': 'CREATE',
#         'has_permission': PermissionChecker(request)
#     }
#
#
# @view_config(
#     route_name='dialog_update_studio',
#     renderer='templates/studio/dialog_create_studio.jinja2'
# )
# def update_studio_dialog(request):
#     """updates the given studio
#     """
#     return {
#         'mode': 'UPDATE',
#         'has_permission': PermissionChecker(request),
#         'studio': Studio.query.first()
#     }


@view_config(
    route_name='create_studio'
)
def create_studio(request):
    """creates the studio
    """
    name = request.params.get('name', None)
    dwh = request.params.get('dwh', None)
    wh_mon_start = get_time(request, 'mon_start')
    wh_mon_end   = get_time(request, 'mon_end')
    wh_tue_start = get_time(request, 'tue_start')
    wh_tue_end   = get_time(request, 'tue_end')
    wh_wed_start = get_time(request, 'wed_start')
    wh_wed_end   = get_time(request, 'wed_end')
    wh_thu_start = get_time(request, 'thu_start')
    wh_thu_end   = get_time(request, 'thu_end')
    wh_fri_start = get_time(request, 'fri_start')
    wh_fri_end   = get_time(request, 'fri_end')
    wh_sat_start = get_time(request, 'sat_start')
    wh_sat_end   = get_time(request, 'sat_end')
    wh_sun_start = get_time(request, 'sun_start')
    wh_sun_end   = get_time(request, 'sun_end')

    if name and dwh:
        # create new studio
        studio = Studio(
            name=name,
            daily_working_hours=int(dwh)
        )
        wh = WorkingHours()

        def set_wh_for_day(day, start, end):
            if start != end:
                wh[day] = [[start.seconds/60, end.seconds/60]]
            else:
                wh[day] = []

        set_wh_for_day('mon', wh_mon_start, wh_mon_end)
        set_wh_for_day('tue', wh_tue_start, wh_tue_end)
        set_wh_for_day('wed', wh_wed_start, wh_wed_end)
        set_wh_for_day('thu', wh_thu_start, wh_thu_end)
        set_wh_for_day('fri', wh_fri_start, wh_fri_end)
        set_wh_for_day('sat', wh_sat_start, wh_sat_end)
        set_wh_for_day('sun', wh_sun_start, wh_sun_end)

        studio.working_hours = wh

        DBSession.add(studio)
        # Commit will be handled by the zope transaction extension

    return HTTPOk()


@view_config(
    route_name='update_studio'
)
def update_studio(request):
    """updates the studio
    """

    studio_id = request.params.get('studio_id')
    studio = Studio.query.filter_by(id=studio_id).first()

    name = request.params.get('name', None)
    dwh = request.params.get('dwh', None)
    wh_mon_start = get_time(request, 'mon_start')
    wh_mon_end   = get_time(request, 'mon_end')
    wh_tue_start = get_time(request, 'tue_start')
    wh_tue_end   = get_time(request, 'tue_end')
    wh_wed_start = get_time(request, 'wed_start')
    wh_wed_end   = get_time(request, 'wed_end')
    wh_thu_start = get_time(request, 'thu_start')
    wh_thu_end   = get_time(request, 'thu_end')
    wh_fri_start = get_time(request, 'fri_start')
    wh_fri_end   = get_time(request, 'fri_end')
    wh_sat_start = get_time(request, 'sat_start')
    wh_sat_end   = get_time(request, 'sat_end')
    wh_sun_start = get_time(request, 'sun_start')
    wh_sun_end   = get_time(request, 'sun_end')

    if studio and name and dwh:
        # update new studio

        studio.name=name
        studio.daily_working_hours=int(dwh)

        wh = WorkingHours()

        def set_wh_for_day(day, start, end):
            if start != end:
                wh[day] = [[start.seconds/60, end.seconds/60]]
            else:
                wh[day] = []

        set_wh_for_day('mon', wh_mon_start, wh_mon_end)
        set_wh_for_day('tue', wh_tue_start, wh_tue_end)
        set_wh_for_day('wed', wh_wed_start, wh_wed_end)
        set_wh_for_day('thu', wh_thu_start, wh_thu_end)
        set_wh_for_day('fri', wh_fri_start, wh_fri_end)
        set_wh_for_day('sat', wh_sat_start, wh_sat_end)
        set_wh_for_day('sun', wh_sun_start, wh_sun_end)

        studio.working_hours = wh

        DBSession.add(studio)
        # Commit will be handled by the zope transaction extension

    return HTTPOk()


@view_config(
    route_name='get_last_schedule_info',
    renderer='json'
)
def get_last_schedule_info(request):
    """returns the last schedule info
    """
    # there should be only one studio
    sql_query = """
    select
        extract(epoch from last_scheduled_at::timestamp AT TIME ZONE 'UTC') * 1000 as last_scheduled_at,
        extract(epoch from last_scheduled_at::timestamp AT TIME ZONE 'UTC' - scheduling_started_at::timestamp AT TIME ZONE 'UTC') as duration,
        "SimpleEntities".name as last_scheduled_by
    from "Studios"
    left outer join "SimpleEntities" on "Studios".last_scheduled_by_id = "SimpleEntities".id
    """

    from stalker import db
    result = db.DBSession.connection().execute(sql_query)
    r = result.fetchone()

    return {
        'last_scheduled_at': r[0],
        'last_scheduling_took': r[1],
        'last_scheduled_by': r[2]
    }
